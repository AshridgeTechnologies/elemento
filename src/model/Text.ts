import Element, {ElementType} from './Element.js'

export default class Text implements Element {
    constructor(
        public id: string,
        public name: string,
        public contentExpr: string) {}

    static is(element: Element): element is Text {
        return element.constructor.name === this.name
    }

    kind: ElementType = 'Text'

}
