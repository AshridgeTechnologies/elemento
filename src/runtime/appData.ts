import create from 'zustand'
import {lensPath, mergeDeepRight, view, set, equals} from 'ramda'


type baseStoreType = {app: {}}
const baseStore = ():baseStoreType => ({app: {}})

const lensFor = (path: string) => lensPath(path.split('.'))

export const useStore = create( () => (baseStore()) )

export const updateState = (path: string, changes: object) => {
    const lens = lensFor(path)
    const existingState = useStore.getState()
    const existingStateAtPath = view(lens, existingState)
    const newStateAtPath = mergeDeepRight(existingStateAtPath, changes)
    if (!equals(newStateAtPath, existingStateAtPath)) {
        const newState = set(lens, newStateAtPath, existingState)
        useStore.setState(newState)
    }
}

export const selectState = (path: string) => (state: object) => view(lensFor(path), state)

export const useObjectState = (path: string) => useStore(selectState(path))

export const _dangerouslyResetState = () => useStore.setState(baseStore(), true)

export const useObjectStateWithDefaults = (path: string, defaults: object) => {
    const existingStateAtPath = useObjectState(path) || {}
    const stateWithDefaults = mergeDeepRight(defaults, existingStateAtPath)
    return stateWithDefaults
}

export const getState = useStore.getState