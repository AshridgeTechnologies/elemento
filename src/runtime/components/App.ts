import React, {createElement} from 'react'
import * as Elemento from '../index'

type Properties = {id: string, pages: { [key: string]: React.FunctionComponent<any>; }}

export default function App({id, pages}: Properties) {
    const defaultPage = Object.keys(pages)[0]
    const state = Elemento.useObjectStateWithDefaults(`${id}._data`, {currentPage: defaultPage})
    const {currentPage} = state
    const pagePath = id + '.' + currentPage

    return createElement('div', {id},
            createElement(pages[currentPage], {path: pagePath}))
}