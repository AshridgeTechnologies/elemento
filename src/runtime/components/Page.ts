import React from 'react'
import {valueOfProps} from '../runtimeFunctions'
import {Stack} from '@mui/material'

type Properties = { path: string, children?: any }

export default function Page({children, path, ...props}: Properties) {
    const propVals = valueOfProps(props)
    return React.createElement(Stack, {
        id: path,
        justifyContent: 'flex-start',
        alignItems: 'flex-start',
        spacing: 2,
        height: '100%', width: '100%',
        sx: {overflowY: 'scroll', padding: 1, position: 'relative'},
        ...propVals,
        children
    })
}
