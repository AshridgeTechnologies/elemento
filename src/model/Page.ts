import Element, {ElementType} from './Element'
import BaseElement,{equalArrays} from './BaseElement'
import Text from './Text'

export default class Page extends BaseElement implements Element {

    constructor(
        id: string,
        name: string,
        private readonly props: {style?: string},
        public readonly elements: ReadonlyArray<Element>
    ) {
        super(id, name)
    }

    kind: ElementType = 'Page'

    static is(element: Element): element is Page {
        return element.constructor.name === this.name
    }

    protected getElements(): ReadonlyArray<Element> { return this.elements }

    get properties() {return this.props}
    get style() { return this.props.style }

    set(id: string, propertyName: string, value: any): Page {
        if (id === this.id) {
            if (propertyName === 'name') {
                return new Page(this.id, value, this.props, this.elements)
            }
            if (propertyName === 'elements') {
                return new Page(this.id, this.name, this.props, value)
            }

            const updatedProps = {...this.props, [propertyName]:value}
            return new Page(this.id, this.name, updatedProps, this.elements)
        }

        const newElements = this.getElements().map( el => el.set(id, propertyName, value))
        if (!equalArrays(newElements, this.getElements())) {
            return new Page(this.id, this.name, this.props, newElements)
        }

        return this
    }

    static fromJSON({id, name, props, elements}: { id: string; name: string; props: any, elements: any[] }): Page {
        return new Page(id, name, props, elements.map(el => Text.fromJSON(el)))
    }
}