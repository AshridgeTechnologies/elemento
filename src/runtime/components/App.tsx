import React, {createElement, useEffect} from 'react'
import {useGetObjectState} from '../appData'
import {Box, Button, Container, Typography} from '@mui/material'
import {SnackbarProvider, enqueueSnackbar, closeSnackbar} from 'notistack'
import {AppData} from './AppData'
import {noop} from '../../util/helpers'
import {isSignedIn, useSignedInState} from './authentication'
import {subscribeToNotifications, type Notification} from './notifications'

import {dndWrappedComponent} from './ComponentHelpers'

type Properties = {path: string, maxWidth?: string | number, fonts?: string[], startupAction?: () => void, children?: any, topChildren?: any}

const containerBoxCss = {
    height: '100%',
    paddingLeft: 2,
    paddingRight: 2,
    paddingTop: 1,
    paddingBottom: 1,
    marginTop: 0,
    marginBottom: 0,
}

const notify = (notification: Notification) => {
    const {level, message, details} = notification
    console.log('Notification', level, message, details)
    const snackbarMessage = <div>
        <Typography color='white'>{message}</Typography>
        <Typography color='white' variant='body2'>{details}</Typography>
    </div>
    enqueueSnackbar(snackbarMessage, {variant: level, persist: level === 'error'})
}

const insertFontLink = (fonts: string[]) => {
    if (fonts.length && !document.head.querySelector('link#web-font-link')) {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        const fontFamilies = fonts.map(font => `family=${font}`).join('&')
        link.href=`https://fonts.googleapis.com/css2?${fontFamilies}&display=swap`
        document.head.appendChild(link)
    }
}

const App: any = dndWrappedComponent(function App({path, maxWidth, fonts = [], startupAction = noop, children, topChildren}: Properties) {
    const state = useGetObjectState<AppData>(path)
    const {currentPage} = state
    const pagePath = path + '.' + currentPage.name

    useEffect( () => { insertFontLink(fonts) }, [] ) // wrap startupAction to ensure no result is returned to useEffect
    useEffect( () => { startupAction() }, [] ) // wrap startupAction to ensure no result is returned to useEffect
    useEffect(() => subscribeToNotifications( notify ), [])
    useSignedInState()
    const pageToDisplay = state.pageToDisplay(isSignedIn())
    return <Box id={path} display='flex' flexDirection='column' height='100%' width='100%'  className='ElApp'>
        <SnackbarProvider action={(snackbarId) => (
            <Button sx={{color: 'white'}} onClick={() => closeSnackbar(snackbarId)}>
                Dismiss
            </Button>
        )}/>
        <Box flex='0'>
            {topChildren}
        </Box>
        <Box flex='1' minHeight={0}>
            <Container maxWidth={false} sx={{...containerBoxCss, maxWidth}}>
                {createElement(pageToDisplay, {path: pagePath} as React.Attributes) /*we do not know the properties of each page*/}
            </Container>
        </Box>
        <Box flex='0'>
            {children}
        </Box>
    </Box>

})

App.State = AppData

export default App
