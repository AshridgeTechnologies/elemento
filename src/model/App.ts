import Page from './Page';

export default class App  {
    constructor(
        public id: string,
        public name: string,
        public pages: Page[]
    ) {}

}