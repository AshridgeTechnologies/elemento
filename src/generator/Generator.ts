import {parse} from 'recast'
import {visit,} from "ast-types"

import App from '../model/App'
import Page from '../model/Page'
import Text from '../model/Text'
import Element from '../model/Element'
import TextInput from '../model/TextInput'
import {globalFunctions} from '../runtime/globalFunctions'
import UnsupportedValueError from '../util/UnsupportedValueError'
import {definedPropertiesOf} from '../util/helpers'

type IdentifierCollector = {add(s: string): void}

function objectLiteral(obj: object) {
    return `{${Object.entries(obj).map(([name, val]) => `${name}: ${val}`).join(', ')}}`
}

const isGlobalFunction = (name: string) => name in globalFunctions

export default class Generator {
    constructor(public app: App) {
    }

    outputFiles() {
        const page = this.app.pages[0]
        return [
            {
                name: `${(page.codeName)}.js`,
                content: Generator.pageContent(page)
            },
            {
                name: 'appMain.js',
                content: Generator.appMainContent(this.app)
            }
        ]
    }

    private static pageContent(page: Page) {
        const stateDefaultEntries = page.elementArray().map( el => Generator.initialStateEntry(el) ).filter( el => !!el )
        const stateDefaultBlock = stateDefaultEntries.length ? `\n        ${stateDefaultEntries.join('\n        ')}\n    `: ''
        const stateDefaultCall = `    const state = useObjectStateWithDefaults(props.path, {${stateDefaultBlock}})`

        const identifierSet = new Set<string>()
        const isPageElement = (name: string) => !!page.elementArray().find( el => el.codeName === name )
        const isKnown = (name: string) => isGlobalFunction(name) || isPageElement(name)
        const pageCode = Generator.generateElement(page, identifierSet, isKnown)
        const identifiers = [...identifierSet.values()]

        const globalFunctionIdentifiers = identifiers.filter(isGlobalFunction)
        const globalDeclarations = globalFunctionIdentifiers.length ? `    const {${globalFunctionIdentifiers.join(', ')}} = window.globalFunctions` : ''
        const pageIdentifiers = identifiers.filter(isPageElement)
        const pageDeclarations = pageIdentifiers.length ? `    const {${pageIdentifiers.join(', ')}} = state` : ''
        const declarations = [globalDeclarations, pageDeclarations].join('\n').trimEnd()


        return `function ${page.codeName}(props) {
    const pathWith = name => props.path + '.' + name
${stateDefaultCall}
${declarations}
    return ${pageCode}
}
`.trimLeft()
    }

    private static appMainContent(app: App) {
        const identifierSet = new Set<string>()
        const page = app.pages[0]
        const identifiers = [...identifierSet.values()]
        const globalFunctionIdentifiers = identifiers.filter( id => isGlobalFunction(id))
        const globalDeclarations = globalFunctionIdentifiers.length ? `    const {${globalFunctionIdentifiers.join(', ')}} = window.globalFunctions` : ''
        return `function AppMain(props) {
${globalDeclarations}
    return React.createElement('div', {id: 'app'},
        React.createElement(${page.codeName}, {path: 'app.${page.codeName}'})
    )
}
`.trimLeft()
    }

    private static generateElement(element: Element, identifiers: IdentifierCollector, isKnown: (name: string) => boolean): string {

        switch(element.kind) {
            case "App":
                const app = element as App
                return this.appMainContent(app)
            case "Page":
                const page = element as Page
                const children = page.elementArray().map(p => `        ${Generator.generateElement(p, identifiers, isKnown)},`).join('\n');
                return `React.createElement('div', {id: props.path},
${children}
    )`
            case "Text":
                const text = element as Text
                const content = Generator.getExprAndIdentifiers(text.contentExpr, identifiers, isKnown)
                return `React.createElement(TextElement, {path: pathWith('${text.codeName}')}, ${content})`

            case "TextInput":
                const textInput = element as TextInput
                const path = `pathWith('${textInput.codeName}')`
                const initialValue = Generator.getExprAndIdentifiers(textInput.initialValue, identifiers, isKnown)
                const maxLength = Generator.getExprAndIdentifiers(textInput.maxLength, identifiers, isKnown)
                const label = Generator.getExprAndIdentifiers(textInput.label, identifiers, isKnown)
                const reactProperties = definedPropertiesOf({path, initialValue, maxLength, label})
                return `React.createElement(TextInput, ${objectLiteral(reactProperties)})`

            default:
                throw new UnsupportedValueError(element.kind)

        }
    }

    private static initialStateEntry(element: Element): string {
        switch(element.kind) {
            case "App":
                const app = element as App
                return ''
            case "Page":
                const page = element as Page
                return ''
            case "Text":
                const text = element as Text
                return ''

            case "TextInput":
                const textInput = element as TextInput
                return `${textInput.codeName}: {value: ${textInput.initialValue || '""'}},`

            default:
                throw new UnsupportedValueError(element.kind)

        }

    }

    private static getExprAndIdentifiers(expr: string | undefined, identifiers: IdentifierCollector, isKnown: (name: string) => boolean) {
        if (expr === undefined) {
            return undefined
        }

        try {
            const ast = parse(expr)
            const thisIdentifiers = new Set<string>()
            visit(ast, {
                visitIdentifier(path) {
                    const node = path.value
                    const parentNode = path.parentPath.value
                    const isPropertyIdentifier = parentNode.type === "MemberExpression" && parentNode.property === node
                    if (!isPropertyIdentifier) {
                        thisIdentifiers.add(node.name)
                    }
                    this.traverse(path);
                }
            })

            const identifierNames = Array.from(thisIdentifiers.values())
            const unknownIdentifiers = identifierNames.filter(id => !isKnown(id))
            if (unknownIdentifiers.length) {
                return `React.createElement('span', {title: "Unknown names: ${unknownIdentifiers.join(', ')}"}, '#ERROR')`
            }

            identifierNames.forEach(name => identifiers.add(name))

            return expr
        } catch(e: any) {
            return `React.createElement('span', {title: "${e.constructor.name}: ${e.message}"}, '#ERROR')`
        }
    }
}