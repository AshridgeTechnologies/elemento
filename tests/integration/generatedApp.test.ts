import renderer from 'react-test-renderer'
import React, {FunctionComponent} from 'react'
import TextElement from '../../src/runtime/TextElement'
import Text from '../../src/model/Text'
import App from '../../src/model/App'
import Generator from '../../src/generator/Generator'
import Page from '../../src/model/Page'
import {useObjectStateWithDefaults} from '../../src/runtime/appData'

let makeComponentFunction = function (functionCode: string) {
    const functionBody = functionCode.replace(/^function \w+.*/m, '').replace(/^}$/m, '').trim()
    return new Function('props', functionBody) as FunctionComponent
}

test('generated Page creates React element with correct structure', ()=> {

    const app = new App('t1', 'test1', {}, [
        new Page('p1', 'Page 1', {}, [
            new Text('id1', 't1', {contentExpr: '"Hi there!"'}),
            new Text('id1', 't2', {contentExpr: '2 + 2'}),
        ])])

    const gen = new Generator(app)
    expect(gen.outputFiles()[0].name).toBe('Page1.js')
    const PageComponent = makeComponentFunction(gen.outputFiles()[0].content)

    global.React = React
    // @ts-ignore
    global.TextElement = TextElement
    // @ts-ignore
    global.useObjectStateWithDefaults = useObjectStateWithDefaults
    // @ts-ignore
    const component = renderer.create(React.createElement(PageComponent, {path: 'app.Page1'}))
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})

test('generated AppMain creates React elements with correct structure', ()=> {

    const app = new App('t1', 'test1', {}, [
        new Page('p1', 'Page 1', {}, [
            new Text('id1', 't1', {contentExpr: '"Hi there!"'}),
            new Text('id1', 't2', {contentExpr: '2 + 2'}),
        ])])

    const gen = new Generator(app)
    expect(gen.outputFiles()[1].name).toBe('appMain.js')

    const Page1 = makeComponentFunction(gen.outputFiles()[0].content)
    const AppMain = makeComponentFunction(gen.outputFiles()[1].content)

    global.React = React
    // @ts-ignore
    global.TextElement = TextElement
    // @ts-ignore
    global.Page1 = Page1
    // @ts-ignore
    global.useObjectStateWithDefaults = useObjectStateWithDefaults
    const component = renderer.create(React.createElement(AppMain))
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})
