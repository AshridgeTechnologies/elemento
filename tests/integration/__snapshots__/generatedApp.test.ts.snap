// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated AppMain creates React elements with correct structure 1`] = `
<div
  id="app"
>
  <div
    id="app.Page1"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
        }
      }
      data-emotion="css d3wcwz-MuiTypography-root"
    />
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
      id="app.Page1.t1"
    >
      Hi there!
    </p>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
        }
      }
      data-emotion="css d3wcwz-MuiTypography-root"
    />
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
      id="app.Page1.t2"
    >
      4
    </p>
  </div>
</div>
`;

exports[`generated Page creates React element with correct structure 1`] = `
<div
  id="app.Page1"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
      }
    }
    data-emotion="css d3wcwz-MuiTypography-root"
  />
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
    id="app.Page1.t1"
  >
    Hi there!
  </p>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
      }
    }
    data-emotion="css d3wcwz-MuiTypography-root"
  />
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
    id="app.Page1.t2"
  >
    4
  </p>
</div>
`;
